### PR TITLE
Guard NumPy usage using USE_NUMPY

### LINKED ISSUE #11757: Unable to build PyTorch from source without NumPy support
## Issue description

So far, all the builds in the CI have had NumPy support, meaning that this issue has not been exposed in its entirety yet.

I tried to build PyTorch without NumPy support, and it wasn't possible at first. Editing `Dependencies.cmake` made sense to accomplish this, but then it seems that Caffe2 requires NumPy support.

This is the error I got after (almost) building PyTorch.
```bash
In file included from /media/vishwak/Official/Projects/pytorch/caffe2/python/pybind_state.cc:1:0:
/media/vishwak/Official/Projects/pytorch/caffe2/python/pybind_state.h:24:10: fatal error: numpy/arrayobject.h: No such file or directory
 #include <numpy/arrayobject.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

## Code example

This is what I did to the `Dependencies.cmake` file:
```diff
diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
index dbec0d07e..c8dafb23e 100644
--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -397,9 +397,12 @@ if(BUILD_PYTHON)
   set(Python_ADDITIONAL_VERSIONS 3.7 3.6 3.5 2.8 2.7 2.6)
   find_package(PythonInterp 2.7)
   find_package(PythonLibs 2.7)
-  find_package(NumPy REQUIRED)
-  if(PYTHONINTERP_FOUND AND PYTHONLIBS_FOUND AND NUMPY_FOUND)
-    include_directories(SYSTEM ${PYTHON_INCLUDE_DIR} ${NUMPY_INCLUDE_DIR})
+  find_package(NumPy)
+  if(PYTHONINTERP_FOUND AND PYTHONLIBS_FOUND)
+    include_directories(SYSTEM ${PYTHON_INCLUDE_DIR})
+    if(NUMPY_FOUND)
+      include_directories(SYSTEM ${NUMPY_INCLUDE_DIR})
+    endif()
     # Observers are required in the python build
     caffe2_update_option(USE_OBSERVERS ON)
   else()
```
## System Info
- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): source
- Build command you used (if compiling from source): `NO_TEST=1 NO_CAFFE2_OPS=1 MAX_JOBS=3 NO_CUDA=1 python setup.py build`
- OS: Ubuntu 18.04.1
- Python version: 3.7
- GCC version (if compiling from source): 7.3
- CMake version: 3.11.3
- Versions of any other relevant libraries: None

### PR DESCRIPTION
**Pull request by vishwakftw**
All usages of the `ndarray` construct have now been guarded with `USE_NUMPY`. This eliminates the requirement of NumPy while building PyTorch from source.

Fixes #11757 

cc: @apaszke

### CODE REVIEW

File: caffe2/python/pybind_state.cc [hunk 1]
```diff
@@ -340,16 +349,19 @@ void addObjectMethods(py::module& m) {
...
+                = reinterpret_cast<PyArrayObject*>(arg.ptr());
               auto feeder = CreateFeeder(option.device_type());
               CAFFE_ENFORCE(
                   feeder, "Unknown device type encountered in FeedBlob.");
               feeder->Feed(option, array, blob);
               return true;
             }
-
+#else
+            CAFFE_THROW("Caffe2 compiled without NumPy support.");
```
Comment (vishwakftw) [inline caffe2/python/pybind_state.cc]: Why is this line causing the Caffe2 failures?

File: caffe2/python/pybind_state.cc [hunk 2]
```diff
@@ -99,9 +100,13 @@ int CaffeToNumpyType(const TypeMeta& meta) {
   };
   const auto it = numpy_type_map.find(meta.id());
   return it == numpy_type_map.end() ? -1 : it->second;
+#else
+  return -1;
```
Comment (apaszke) [inline caffe2/python/pybind_state.cc]: This should throw

File: caffe2/python/pybind_state.cc [hunk 3]
```diff
@@ -125,6 +130,10 @@ const TypeMeta& NumpyTypeToCaffe(int numpy_type) {
   static TypeMeta unknown_type;
   const auto it = caffe_type_map.find(numpy_type);
   return it == caffe_type_map.end() ? unknown_type : it->second;
+#else
+  static TypeMeta unknown_type;
+  return unknown_type;
```
Comment (apaszke) [inline caffe2/python/pybind_state.cc]: This should throw

**Review by apaszke — COMMENTED** [review]
cc @Yangqing for review

**Comment by Yangqing** [comment]
So, regarding the test failures you encounter, it is because the -DUSE_NUMPY is only passed to cmake but not passed down to the compilers. To do so, you should add 

#cmakedefine USE_NUMPY

in caffe2/core/macros.h.in and you should be able to properly get the with-numpy version. Thanks so much for the rest of the PR! It looks great.

**Comment by Yangqing** [comment]
By the way, when importing this into Facebook internal code, we will need to add the USE_NUMPY macro in our internal build as well, so I'll self-assign this PR to myself to import.

**Comment by vishwakftw** [comment]
Thank you @Yangqing for the solution. I think I've done the necessary as advised, the tests should pass now I suppose.

**Comment by vishwakftw** [comment]
The test failures persist, originating from the same place as earlier.

**Comment by vishwakftw** [comment]
All tests pass now, the test failures are related to DDP.

**Comment by vishwakftw** [comment]
@apaszke could I get a review on this?

**Comment by vishwakftw** [comment]
@Yangqing could I get a review on this?

**Comment by vishwakftw** [comment]
Now all the builds have passed. I think the fix is right.

**Comment by SsnL** [comment]
@Yangqing @apaszke Is this good to go?

File: caffe2/python/pybind_state.h [hunk 4]
```diff
@@ -161,9 +174,10 @@ class TensorFeeder : public BlobFeederBase {
  public:
   void FeedTensor(
       const DeviceOption& option,
-      PyArrayObject* original_array,
+      PyObject* original_array,
```
Comment (apaszke) [inline caffe2/python/pybind_state.h]: Hmm can we keep it as a `PyArrayObject*` but forward declare `struct PyArrayObject;` if `USE_NUMPY` is false? It's really much safer.

**Review by apaszke — COMMENTED** [review]
Can we please revert the `PyArrayObject -> PyObject` change?

File: caffe2/python/pybind_state.h (hunk 4 above)
Comment (vishwakftw) [reply caffe2/python/pybind_state.h]: Sure, I will do that. I had thoughts about this earlier, but then I looked at `torch/csrc/utils/tensor_numpy.*` and followed the convention there.

Comment (vishwakftw) [reply caffe2/python/pybind_state.h]: @apaszke this is fixed.

**Review by apaszke — APPROVED** [review]
LGTM, but @Yangqing might want to stamp this too

**Comment by vishwakftw** [comment]
@Yangqing is this good to go?

**Comment by SsnL** [comment]
extra places to be guarded:
https://github.com/pytorch/pytorch/blob/a2ebbccc9f04abc89c37e3227355858d668ac000/caffe2/python/pybind_state_mkl.cc#L33
https://github.com/pytorch/pytorch/blob/a2ebbccc9f04abc89c37e3227355858d668ac000/caffe2/python/pybind_state_mkl.cc#L72
https://github.com/pytorch/pytorch/blob/a2ebbccc9f04abc89c37e3227355858d668ac000/caffe2/python/pybind_state_int8.cc#L39

**Comment by SsnL** [comment]
@apaszke @Yangqing This patch somehow makes internal caffe2 compiled **without** numpy support... See the fbcode tests failures.

**Comment by noname2534** [comment]
ในวันที่ อา. 30 ก.ย. 2018 14:37 น. Tongzhou Wang <notifications@github.com>
เขียนว่า:

**Comment by SsnL** [comment]
@vishwakftw I talked with yangqing and I'm patching internal build right now.

**Comment by vishwakftw** [comment]
Thank you @SsnL for working on this.

**Pull request closed without merging**